### PR TITLE
Implemented density transfer, added timings to hessian

### PR DESCRIPTION
**Pull request by a-azzali**
Added `transfer_density`
Added a non-k-point-specific `transfer_mapping`
Added timings to hessian functions

### CODE REVIEW

File: src/transfer.jl [hunk 1]
```diff
@@ -122,6 +150,32 @@ function transfer_blochwave(ψ_in, basis_in::PlaneWaveBasis{T},
...
+"""
+function transfer_density(ρ_in, basis_in::PlaneWaveBasis{T},
+		          basis_out::PlaneWaveBasis{T}) where {T}
+
+    @assert basis_in.model.lattice == basis_out.model.lattice
+    @assert length(size(ρ_in)) ∈ (3, 4)
+
+    ρ_freq_in = fft(basis_in, ρ_in)
+    ρ_freq_out = similar(ρ_freq_in, basis_out.fft_size..., size(ρ_in, 4))
+    ρ_freq_out .= 0
```
Comment (antoine-levitt) [inline src/transfer.jl:168]: We have a `zeros_like` for this

File: src/transfer.jl [hunk 2]
```diff
@@ -122,6 +150,32 @@ function transfer_blochwave(ψ_in, basis_in::PlaneWaveBasis{T},
...
+
+    idcs_in, idcs_out = transfer_mapping(basis_in, basis_out)
+    map(zip(idcs_in, idcs_out)) do (block_in,block_out)
```
Comment (antoine-levitt) [inline src/transfer.jl:171]: why not do this with a plain old loop?

File: src/transfer.jl [hunk 3]
```diff
@@ -122,6 +150,32 @@ function transfer_blochwave(ψ_in, basis_in::PlaneWaveBasis{T},
...
+        ρ_freq_out[block_out,:] .= ρ_freq_in[block_in,:]
+    end
+    
+    ρ_real_out = ifft(basis_out, ρ_freq_out)
+    ρ_real_out
```
Comment (antoine-levitt) [inline src/transfer.jl:176]: no need for that last line

**Review by antoine-levitt — COMMENTED** [review]
Hi and thanks for this, looks useful! Generally the code looks a bit complicated because of all the cartesian indices. Would it not be simpler to vectorize the arrays (ie rho_vec = reshape(rho, prod(fft_size), 2) and have transfer_mapping work with just integers?

**Comment by a-azzali** [comment]
Thank you for suggesting those 3 changes to `transfer_density`, I hope it's ok now.

About the array vectorization: We're using the CartesianIndices because one has 8 parallelepiped-shaped blocks of indices to be returned (we wanted to make use of this). I haven't found an easy way of doing this in 1d...

Btw I should have implemented the tests for the density transfer and fixed some bugs.

**Pull request merged**